File: Medium/Blackjack (with classes)/Python/blackjack.py
import random
class Card:
    values = ["Ace", "1", "2", "3", "4", "5", "6", "7", "8", "9", "10", "Jack", "Queen", "King"]
    suits = ["Diamonds", "Hearts", "Clubs", "Spades"]
    def __init__(self, value, suit):
        if value not in self.values: 
            raise ValueError(value + " is not a valid value")
        if suit not in self.suits:
            raise ValueError(suit + " is not a valid suit")
        self.value = value
        self.suit = suit

    def __repr__(self):
        return self.value + " of " + self.suit
    
    def int_value(self):
        if self.value == "Ace":
            return 11
        elif self.value in ["Jack", "Queen", "King"]:
            return 10
        else:
            return int(self.value)
    
# build the deck
deck = []
for value in Card.values:
    for suit in Card.suits:
        new_card = Card(value, suit)
        deck.append(new_card)

random.shuffle(deck)

def hand_total(hand):
    total = 0
    hasAce = False
    for card in hand:
        if card.value == "Ace":
            hasAce = True
        total += card.int_value()
    if hasAce and total > 21:
        total -= 10
    return total 

player_hand = []
dealer_hand = []

for i in range(2):
    player_hand.append(deck.pop())
    dealer_hand.append(deck.pop())

print("Welcome to blackjack!")
print("The value of an ACE automatically changes to whatever benefits you the most")
print("The dealer's first card is a", dealer_hand[0])

# iterate through the player's turn 
while True:
    print("---")
    print("Your current hand is", player_hand)
    player_hand_total = hand_total(player_hand) 
    if player_hand_total >= 21:
        break
    print("Your hand's total is", player_hand_total)
    answer = input("Would you like to hit or stand?: ").lower()
    if answer == "hit":
        new_card = deck.pop()
        print("You drew a", new_card)
        player_hand.append(new_card)
    elif answer == "stand":
        print("You chose to stand")
        break
    else:
        print("Invalid option")
print("-----")
player_hand_total = hand_total(player_hand)
print("Your final total is:", hand_total(player_hand))

# the := operator (walrus operator) is used to reassign the value of dealer_hand_total AND compare it to 16, in the same line 
# if you don't like it, you're not alone. the creator of python got so much criticism for adding it that he stepped down!
while (dealer_hand_total := hand_total(dealer_hand)) <= 16:
    dealer_hand.append(deck.pop())
print("-----")
print("The dealer's final hand is", dealer_hand)
print("The dealer's final total is", dealer_hand_total)
if player_hand_total == dealer_hand_total or ((player_hand_total > 21) and (dealer_hand_total > 21)):
    print("It was a tie")
elif player_hand_total <= 21 and (player_hand_total > dealer_hand_total or dealer_hand_total > 21):
    print("You won!")
else:
    print("You lost. . .")

File: Medium/Blackjack (with classes)/Java/Blackjack.java
import java.util.List;
import java.util.ArrayList;
import java.util.Random;
import java.util.Scanner;

public class Blackjack {
    public static void main(String[] args) throws Exception {
        // fill the deck with cards
        List<Card> deck = new ArrayList<>();
        for (String value : Card.Values) {
            for (String suit : Card.Suits) {
                Card newCard = new Card(value, suit);
                deck.add(newCard);
            }
        }
        List<Card> playerHand = new ArrayList<>();
        List<Card> dealerHand = new ArrayList<>();

        for (int i = 0; i < 2; i++) {
            Card newCard1 = drawCard(deck);
            playerHand.add(newCard1);
            Card newCard2 = drawCard(deck);
            dealerHand.add(newCard2);

        }
        Scanner input = new Scanner(System.in);
        System.out.println("Welcome to blackjack!");
        System.out.println("The value of an ACE automatically changes to whatever benefits you most.");
        System.out.println("The dealer's first card is a: " + dealerHand.getFirst());

        // iterate through the player's turn
        while (true) {
            System.out.println("---");
            System.out.print("Your current hand is: ");
            printHand(playerHand);
            int handTotal = getHandTotal(playerHand);
            if (handTotal >= 21) {
                break;
            }
            System.out.println("Your hand's total is " + handTotal);
            System.out.print("\nWould you like to hit or stand?: ");
            String answer = input.nextLine().toLowerCase();
            if (answer.equals("hit")) {
                Card newCard = drawCard(deck);
                System.out.println("You drew a " + newCard);
                playerHand.add(newCard);
            }
            else if (answer.equals("stand")) {
                System.out.println("You choose to stand.");
                break;
            }
            else {
                System.out.println("Invalid option!");
            }
        }
        int playerFinalTotal = getHandTotal(playerHand);
        System.out.println("-----");
        System.out.println("Your final total is: " + playerFinalTotal);
        // iterate through the dealer's turn
        // dealer stops once hand total is more than 16
        int dealerFinalTotal;
        while ((dealerFinalTotal = getHandTotal(dealerHand)) <= 16) {
            dealerHand.add(drawCard(deck));
        }
        System.out.println("-----");
        System.out.print("The dealer's final hand is: ");
        printHand(dealerHand);
        System.out.println("The dealer's final total is: " + dealerFinalTotal);
        if ((playerFinalTotal == dealerFinalTotal) || (playerFinalTotal > 21 && dealerFinalTotal > 21)) {
            System.out.println("It was a tie!");
        }
        else if (playerFinalTotal <= 21 && (playerFinalTotal > dealerFinalTotal || dealerFinalTotal > 21)) {
            System.out.println("You won!");
        } else {
            System.out.println("You lost. . . ");
        }
        input.close();
    }
    public static void printHand(List<Card> hand) {
        System.out.print("[  ");
        for (Card card : hand) {
            System.out.print(card + "  ");
        }
        System.out.println("]");
    }
    public static Card drawCard(List<Card> deck) {
        Random random = new Random();
        return deck.remove(random.nextInt(deck.size()));
    }
    public static int getHandTotal(List<Card> hand) {
        boolean hasAce = false;
        int total = 0;
        for (Card card : hand) {
            if (card.value().equals("Ace")) {
                hasAce = true;
            }
            total += card.getIntValue();
        }
        // Aces are 11 by default but switch to 1 if the total exceeds 21.
        if (total > 21 && hasAce) {
            return total - 10;
        }
        return total;
    }
}
